apps/design/frontend/src/rich_text_editor.test.tsx

```diff
@@ -0,0 +1,119 @@
+import userEvent from '@testing-library/user-event';
+import { fireEvent, render, screen } from '../test/react_testing_library';
+import { RichTextEditor } from './rich_text_editor';
+
+// userEvent.type only works when the editor is empty for some reason - this
+// appends to existing content
+function type(editor: HTMLElement, text: string) {
+  fireEvent.change(editor.querySelector('.tiptap')!, {
+    target: {
+      textContent: editor.querySelector('.tiptap p')!.textContent + text,
+    },
+  });
+}
```

Comment:
I looked into why this happens. It seems that versions of `@testing-library/user-event` prior to 14.0.0 crashed when attempting to type into a `[contenteditable]` element that was not empty due to a bug with trying to set the selection. It would throw a `DOMException` in `jsdom` like so:

```ts
  if (offset > nodeLength(node)) {
    throw DOMException.create(node._globalObject, ["Offset out of bound.", "IndexSizeError"]);
  }
```

This happened because `offset` was the length of the text content of the `[contenteditable]` element (i.e. 7 when the text content is "Content") and `nodeLength(node)` is computed as the child count of `node` (i.e. 1 when the content of the `[contenteditable]` element is `<p>Content</p>`).

This workaround happens to poke `tiptap` just the right way to make it do what we want, but it feels kinda brittle. I investigated whether a newer version of `@testing-library/user-event` fixes this issue and it _sort of_ does. In https://github.com/testing-library/user-event/pull/822, basic support for updates to `[contenteditable]` elements is added and released in v14.0.0. However, when I updated to v14.5.2 (latest) and changed to `userEvent.type` in the "bold formatting" test below, it failed because the content ended up as `<p><strong> bold</strong>Content</p>`. This is because the selection defaults to the beginning of the content area. I tried using `userEvent.type(editor, ' bold', { initialSelectionStart: 7 })` or `userEvent.type(editor, '[End]')` beforehand, but in each case I get [this error](https://github.com/testing-library/user-event/blob/d0362796a33c2d39713998f82ae309020c37b385/src/event/selection/setSelectionRange.ts#L33):

> Not implemented. The result of this interaction is unreliable.

This happens because `@testing-library/user-event` only sets selection ranges for `input`, `textarea`, and `[contenteditable]` where there is a single child node and it's just text.

In summary, `@testing-library/user-event` does not appear to support testing with `[contenteditable]` in anything but the most basic way. I suggest we replace most or all of the tests in this file with Playwright tests.

Reply:
Thanks for looking into this. I agree that the workaround is a bit brittle and that Playwright tests would be a better overall solution. I briefly considered switching to Playwright tests when I first realized `user-event` wasn't working. I want to stick with this for now because it's working and I don't want to put more time into it, given that VxDesign is still an internal tool and won't be part of the upcoming release. How does that sound to you?

Reply:
I don't have a problem with that, but I'd suggest adding a comment with a link to a Github issue to change to Playwright. Unless you think we _shouldn't_ change it in the future?

Reply:
Sounds good: e69a916a4